### PR TITLE
Experimental Title Screen port

### PR DESCRIPTION
**Pull request by NextinMono**
Title Screen port. Keep in mind that it's not finished, and lacks these following features:
-It doesn't move to the Continue button with a save present
-Movement is still vertical and not horizontal
-No music

Also has title screen background from Exe.Qutionn (with permission), and has no World Map.

Should probably also check for the code, if it's consistent and if it's good, you guys know better than me about this stuff!

### CODE REVIEW

**Comment by NextinMono** [comment]
(also temporarily, the textures of the main menu havent been touched, so online mode and quit game will not say their respective names)

**Comment by hyperbx** [comment]
Rename HudTitle to just Title, the classes are supposed to follow the naming conventions from Generations itself, just without the C prefix.

File: DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp [hunk 1]
```diff
@@ -0,0 +1,183 @@
+#include "HudResult.h"
```
Comment (hyperbx) [inline DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp]: Remove these includes, they're already in the pre-compiled header.

File: DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+#include "HudSonicStage.h"
+#include "HudLoading.h"
+#include "HudTitle.h"
+
+
```
Comment (hyperbx) [inline DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp]: Remove these line breaks.

File: DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+int currentTitleIndex = 0;
+
+void HudTitle::CreateScreen(Sonic::CGameObject* pParentGameObject)
+{
+	if (rcTitleScreen && !spTitleScreen)
+		pParentGameObject->m_pMember->m_pGameDocument->AddGameObject(spTitleScreen = boost::make_shared<Sonic::CGameObjectCSD>(rcTitleScreen, 0.5f, "HUD", false), "main", pParentGameObject);
+}
+void HudTitle::KillScreen()
```
Comment (hyperbx) [inline DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp]: Please space out your functions with a line break.

File: DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+	printf("CTitlePressSelectIntro");
+	return originalCTitlePressSelectIntro(This);
+}
+HOOK(int, __fastcall, CTitleInit, 0x11D12A0, int a1)
+{
+	printf("CTitleInit");
+	return originalCTitleInit(a1);
+}
+
+
```
Comment (hyperbx) [inline DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp]: Remove two of these line breaks.

File: DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/Mod.cpp [hunk 5]
```diff
@@ -1,4 +1,4 @@
-
+#include "HudTitle.h"
```
Comment (hyperbx) [inline DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/Mod.cpp]: Remove this and include it in `Pch.h` instead.

File: DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.h [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+public:
+	static void Install();
+
+	static void CreateScreen(Sonic::CGameObject* pParentGameObject);
+	static void ToggleScreen(const bool visible, Sonic::CGameObject* pParentGameObject);
+	static void KillScreen();
+	static void FreezeMotion(Chao::CSD::CScene* pScene);
+	static void PlayAnimation(Chao::CSD::CScene* pScene, const char* name, Chao::CSD::EMotionRepeatType repeatType, float motionSpeed, float startFrame);
+	static void IntroAnim(Chao::CSD::RCPtr<Chao::CSD::CScene> scene);
+
```
Comment (hyperbx) [inline DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.h]: Remove this line break.

File: DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+	if (inputPtr->LeftStickVertical >= 0.5f)
+	{
+		currentTitleIndex -= 1;
+		HudTitle::PlayAnimation(*rcTitleMenu, "Scroll_Anim_2_2", Chao::CSD::eMotionRepeatType_PlayOnce, 1, 0);
+		UnleashedTitleText();
+	}
+	if (inputPtr->LeftStickVertical < -0.5f)
+	{
+		currentTitleIndex += 1;
+
```
Comment (hyperbx) [inline DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp]: Line break inconsistent with the condition above it.

File: DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+	printf("%d", inputPtr->LeftStickHorizontal);
+
+
+	if (inputPtr->LeftStickVertical >= 0.5f)
+	{
+		currentTitleIndex -= 1;
+		HudTitle::PlayAnimation(*rcTitleMenu, "Scroll_Anim_2_2", Chao::CSD::eMotionRepeatType_PlayOnce, 1, 0);
+		UnleashedTitleText();
+	}
+	if (inputPtr->LeftStickVertical < -0.5f)
```
Comment (hyperbx) [inline DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp]: Add a line break above this if statement.

File: DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+		spTitleScreen->SendMessage(spTitleScreen->m_ActorID, boost::make_shared<Sonic::Message::MsgKill>());
+		spTitleScreen = nullptr;
+	}
+}
+
+void HudTitle::ToggleScreen(const bool visible, Sonic::CGameObject* pParentGameObject)
+{
+	if (visible)
+	{
+		CreateScreen(pParentGameObject);
```
Comment (hyperbx) [inline DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp]: Remove the braces for this if statement.

File: DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+		currentTitleIndex = 4;
+	if (currentTitleIndex > 4)
+		currentTitleIndex = 0;
+
+	rcTitleMenuTXT->GetNode("txt_0")->SetHideFlag(!(currentTitleIndex == 0));
+	rcTitleMenuTXT->GetNode("txt_1")->SetHideFlag(!(currentTitleIndex == 1));
+	rcTitleMenuTXT->GetNode("txt_2")->SetHideFlag(!(currentTitleIndex == 2));
+	rcTitleMenuTXT->GetNode("txt_3")->SetHideFlag(!(currentTitleIndex == 3));
+
+	printf("\n CURRENT INDEX: %d", currentTitleIndex);
```
Comment (hyperbx) [inline DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp]: Remove your debug output or wrap it with pre-processor flags for debug builds only.

Example;
```
#if _DEBUG
// debug code we don't want in Release
#endif
```

File: DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+	return originalCTitleMainWait(a1);
+}
+
+HOOK(int, __fastcall, CTitleMainFinish, 0x5727F0, Sonic::CGameObject* This)
+{
+	auto inputState = Sonic::CInputState::GetInstance();
+	auto inputPtr = &inputState->m_PadStates[inputState->m_CurrentPadStateIndex];
+	printf("%d", inputPtr->LeftStickHorizontal);
+
+
```
Comment (hyperbx) [inline DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp]: Remove this line break.

**Review by hyperbx — CHANGES_REQUESTED** [review]

File: DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/Mod.cpp [hunk 12]
```diff
@@ -18,6 +18,7 @@ extern "C" __declspec(dllexport) void Init(const char* path)
 	HudLoading::Install();
 	HudPause::Install();
 	Patches::Install();
+	HudTitle::Install();
```
Comment (hyperbx) [inline DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/Mod.cpp]: Move this above `Patches::Install();` for consistency.

**Comment by NextinMono** [comment]
Thank you for fixing the stuff related to consistency, as for debugging stuff, yeah I could've defo used a preprocessor but I forgot that they existed

File: DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+	rcTitleMenuScroll = rcTitleScreen->CreateScene("menu_scroll");
+	rcTitleMenuTXT = rcTitleScreen->CreateScene("txt");
+	HudTitle::FreezeMotion(*rcTitleMenuTXT);
+	rcTitleMenuTXT->GetNode("txt_4")->SetHideFlag(true);
+
+	HudTitle::FreezeMotion(*rcTitleMenuScroll);
+	HudTitle::FreezeMotion(*rcTitleMenu);
+	HudTitle::FreezeMotion(*rcTitlebg);
+
+
```
Comment (hyperbx) [inline DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp]: Remove this line break.

File: DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+	scene->m_MotionSpeed = 1;
+	scene->m_MotionRepeatType = Chao::CSD::eMotionRepeatType_PlayOnce;
+	scene->Update(0.0f);
+}
+void UnleashedTitleText()
+{
+	rcTitleMenuTXT->SetPosition(0, 0);
+	if (currentTitleIndex < 0)
+		currentTitleIndex = 4;
+	if (currentTitleIndex > 4)
```
Comment (hyperbx) [inline DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp]: Make this an `else if` and put a line break above the main if statement.

**Review by hyperbx — CHANGES_REQUESTED** [review]

File: DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp (hunk 1 above)
Comment (NextinMono) [reply DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp]: You mean all of them or only HudResult?

Comment (hyperbx) [reply DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp]: All of them, see this code review as to why: https://github.com/brianuuu/PTKickass_ModUpdates/pull/1/files#r889574673

File: DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+	scene->Update(0.0f);
+}
+
+void UnleashedTitleText()
+{
+	rcTitleMenuTXT->SetPosition(0, 0);
+
+	if (currentTitleIndex < 0)
+		currentTitleIndex = 4;
+	else if(currentTitleIndex > 4)
```
Comment (hyperbx) [inline DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp]: Put a space after `if`.

Comment (NextinMono) [reply DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp]: You mean an extra space after the else if to fit with the if?

Comment (hyperbx) [reply DLLModsSource/SonicGenerations/SonicUnleashedHUD/UnleashedHUD/HudTitle.cpp]: Yes.

**Comment by brianuuu** [comment]
I feel like you should make this as it's own mod, the biggest problem right now I think is it's not "complete", like you mentioned there are bugs that are still presents and I'm probably not gonna fix them for you as title screen isn't the main focus of this mod. But if you want specific sound effects present in Unleashed HUD you can always make this mod as a dependency.

**Comment by NextinMono** [comment]
Yeah I can make it a separate mod, I wanted to include it in this since it would've been nicer if all of the UI ports were a single mod, but it's not a big deal. Thank you!

**Pull request closed without merging**